_posts/2017-09-12-repeatable-serverless-applications.markdown

```diff
@@ -0,0 +1,108 @@
+
+```bash
+if [ $ENV = "local" ]; then
+    aws --endpoint-url=http://localhost:4572 s3 mb s3://instance-reaper-$COMPANY_NAME-logging
+else
+    aws s3 mb s3://instance-reaper-$COMPANY_NAME-logging --region $REGION
+fi
+```
+
+You can now test code locally, that can send and receive data from a dynamoDB table and leverage s3 buckets for file storage.
+We were using the boto3 library provided by AWS, which allows you to set the endpoint URL of the service you want to talk to when creating a client connection to it. Using this endpoint URL parameter, we were able to establish a connection to our s3 bucket on LocalStack, when the 'ENV' environment variable was set to 'local'. This way, we don't get charged for creating excess and unnecessary resources on our AWS account. 
+
+```python
+s3_endpoint = "http://localhost:4572" if os.environ.get("ENV") == "local" else None
+self.s3_resource = boto3.resource('s3', endpoint_url=s3_endpoint)
+```
+
+If you're hoping to use DynamoDB as a layer of persistence in your application and you are using Python, you may want to use the PynamoDB library which will create a table, if it doesn't already exist. To establish a connection to your local instance of Dynamo, you will have to specify the host name, in the Meta subclass, to point to Dynamo on LocalStack.
+
+```python
+class ProductModel(Model):
+    class Meta:
+        table_name = 'Products'
+        region = 'ap-southeast-2'
+        host = "http://localhost:4569" if os.environ.get(
+            "ENV") == "local" else None 
+```
+
+Since LocalStack is a docker image, with all the AWS services you would typically use, in a cloud native application, you can assume that your application would run the same locally, as it would in the cloud. Subsequently, you should test your application assuming that all the application necessary services are present, when developing locally.  
+
+```python
+def test_write_data():
+    ''' test the write_data function, which writes over
+     50 items to our DynamoDB table.'''
+    DataRepository().write_data()
+    # check that our table exists
+    assert ProductModel.exists() is True
+    # perform a "scan" operation on our table
+    for item in ProductModel.scan(limit=50):
+        # checks that we can read at least 50 items
+        # Perform a type check on the object received from the scan
+        if isinstance(item, ProductModel):
+            break
+```
+
+Testing on a pipeline is also a relatively trivial endeavour. Using BitBucket pipelines, we are able to attach the LocalStack docker image, which includes all the necessary AWS services for the lifetime of our pipeline. No additional configuration is required, simply run the same tests you did locally, on your pipeline. You can now use the outcome of your pipeline, as a deciding factor, to whether you will deploy your application to AWS.
```

Comment:
`Testing on a pipeline is also a relatively trivial endeavour`
This assumes your pipeline supports containers.